processor/src/test/java/org/mapstruct/ap/test/bugs/_1576/java8/Source.java

```diff
@@ -0,0 +1,21 @@
+/*
+ * Copyright MapStruct Authors.
+ *
+ * Licensed under the Apache License version 2.0, available at http://www.apache.org/licenses/LICENSE-2.0
+ */
+package org.mapstruct.ap.test.bugs._1576.java8;
+
+import java.util.Date;
+
+public class Source {
```

Comment:
> The problem is indeed that calling `TypeFactory#getType` marks the entries as imported. Maybe we need to distinguish between types that are used in the generation and types that aren't

I thought about that. But i still think its to brittle. In principle all types created by the `getType` can be used later on. We simply don't know at forehand. We need the functionality in type, because we check on `assignment` than on the FQN (need to check this still).

The problem is that determining whether a type is imported belongs to the `GeneratedType` and not as such to `Type`. We need it earlier, during model creation (checkout `Type#getReferenceName`) and we only 'know' which types we already added during writing the model. Its the famous catch22. 

------------

The method name `TypeFactory#getType` is misleading.. Its actually `createType`, which is a bit strange at the same time, because it might actually already have been created by another mapping method (perhaps we should cache already created types in the future, but that's a different optimisation and return a cached one in the `getType` in the future).

Reply:
Thought a little bit more about it.. There are 2 moments when we need to know if a type is imported:
1. When its referenced (during creation: `Type#getReferenceName`)
2. When we generate the imports for the `GeneratedType`.

What if we move the administration whether a type is imported or not to this `Type#getReferenceName`? We could rebrand it as `Type#createReference` returning the same `String` but additionally doing the administration that is now in the `TypeFactory#isImported`..

Than we don't have to be so picky when to call the `TypeFactory#getType`. WDYT?

Reply:
Regarding the caching it has come to my mind as well, but I would say it should be done some other time. We should create some benchmarking to see if we gain some speed from this.

Regarding the reference and the import. I agree with you. I think that we can even leave the `getReferenceName` and always delegate to `TypeFactory#isImported` to check if a type is imported and not have a boolean in the `Type`. This would mean that we need to have the `TypeFactory` in the `Type`, but I think this is OK. This would mean that the last commit is obsolete, right?

Reply:
Apart from the test case, yes. I probably only delegate v the hash table